witness/src/Dscp/Witness/Lock.hs

```diff
@@ -0,0 +1,63 @@
+
+module Dscp.Witness.Lock
```

Comment:
I doubt this is a good design decision.
* We might have more than one lock, so it makes sense to abstract the functionality
* Because of that, the name `Lock` is too generic
* `readingLockOf` and `writingLockOf` are really vague names for the functions, what you really mean is `withReadingLock`
* I do not like the "ing", can't it be something like `takeSharedLock` or `takeLockRead`? This one is subjective, of course.